File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        p = 1
        n = len(nums)
        output = []
        #first build left product.
        for i in range(0,n):
            output.append(p)
            p = p * nums[i]
        p = 1
        # perform right product
        for i in range(n-1, -1, -1):
            output[i] = output[i] * p
            p = p * nums[i]
        return output
    
'''
dry run 
input
1 2 3 4 5 

p value 1 1 2 6 24 
output  1 1 2 6 24

p value 120
output  120 60 40 30 24
'''

File: 238-product-of-array-except-self/238-product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        int pro = 1, zero_cnt = 0, n = nums.size();
        for(int i = 0; i < n; i++){
            if(nums[i] != 0) pro *= nums[i];
            else zero_cnt++;
        }
        vector<int> ans(n, 0);
        if(zero_cnt > 1){
            return ans;
        }
        for(int i = 0; i < n; i++){
            if(nums[i] != 0 && zero_cnt == 0){
                ans[i] = pro/nums[i];
            }else if(nums[i] == 0 && zero_cnt == 1){
                ans[i] = pro;
            }
        }
        return ans;
    }
};